### PR TITLE
Add example blueprint for CSI snapshot functions

### PR DESCRIPTION
**Pull request by shlokchaudhari9**
## Change Overview

This PR consists of Blueprint example and usage of CSI Snapshot Functions to backup/restore MySQL DB

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [x] :world_map: Documentation
- [ ] :robot: Test

## Issues

- #1126 

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: examples/stable/mysql-csi-snapshot/README.md [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+> **Note**: Please use appropriate snapshotClass and storageClass in the ./mysql-csi-snapshot-bp.yaml file. You can correct the blueprint using kubectl patch or kubectl edit commands.
+
+```bash
+$ kubectl create -f ./mysql-csi-snapshot-bp.yaml -n kanister
+```
+
+## Backup the application data
+
+Take a backup of the MySQL data using the backup ActionSet from above blueprint. Create an ActionSet in the same namespace as the controller. An easy way to do this is by using ``kanctl``.
```
Comment (kale-amruta) [inline examples/stable/mysql-csi-snapshot/README.md:99]:
```suggestion
Take a backup of the MySQL data using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+backup-mlvcv                 112s
+
+# View the status of the actionset
+# Please make sure the name of the actionset here matches with name of the name of actionset that we have created already
+$ kubectl --namespace kanister describe actionset backup-mlvcv
+
+# Check the CSI VolumeSnapshot created
+$ kubectl -n mysql get volumesnapshot
+NAME                                  READYTOUSE   SOURCEPVC              SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS      SNAPSHOTCONTENT                                    CREATIONTIME   AGE
+data-mysql-release-0-snapshot-cd26z   true         data-mysql-release-0                           8Gi           do-block-storage   snapcontent-2e411d1d-0b5f-48d4-9b79-b19e824c2e38   3h40m          3h40m
```
Comment (kale-amruta) [inline examples/stable/mysql-csi-snapshot/README.md:118]: Mention that you are using DO cluster hence the snapshhotclass is `do-block-storage`

File: examples/stable/mysql-csi-snapshot/README.md [hunk 3]
```diff
@@ -0,0 +1,267 @@
+# MySQL DB using CSI VolumeSnapshots
```
Comment (kale-amruta) [inline examples/stable/mysql-csi-snapshot/README.md:1]: Title doesnt specify that we are doing backup/restore

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/README.md:1]: Should we set the title as "Backup and restore MySQL DB using CSI VolumeSnapshot"?

File: examples/stable/mysql-csi-snapshot/README.md [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+# Backup MySQL DB
+
+## Create Blueprint
+
+Create Blueprint in the same namespace as the Kanister controller.
+
+> **Note**: We used a Kubernetes cluster on DigitalOcean. Hence, snapshotClass and storageClass in the ./mysql-csi-snapshot-bp.yaml file is set to `do-block-storage`. Please correct these arguments as per your cluster setup. You can make use of `kubectl patch` or `kubectl edit` commands to make these changes.
+
+```bash
+$ kubectl create -f ./mysql-csi-snapshot-bp.yaml -n kanister
```
Comment (oliverblaha) [inline examples/stable/mysql-csi-snapshot/README.md:94]: file is named mysql-csi-snapshot-**blueprint**.yaml

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/README.md:94]: Oh. My bad. I missed that.

File: examples/stable/mysql-csi-snapshot/README.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
+# Backup and restore MySQL DB using CSI VolumeSnapshot
+
+VolumeSnapshots provide Kubernetes users with a standardized way to copy a volume's contents at a particular point in time without creating an entirely new volume. This functionality enables, for example, database administrators to backup databases before performing edit or delete modifications.
+
+## Introduction
+This document explains how Kanister leverages the use of CSI VolumeSnapshots to take backup and restore of databases in MySQL.
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:6]:
```suggestion
This document explains how Kanister leverages the use of CSI VolumeSnapshots to snapshot and restore MySQL database.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+## Prerequisites
+
+- Helm 3 installed
+- Kubernetes 1.16+ with Beta APIs enabled
+- PV provisioner support in the underlying infrastructure
+- Kanister controller version 0.71.0 installed in your cluster, let's assume in Namespace `kanister`
+- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
+- CSI Driver installed
+- VolumeSnapshotClass and StorageClass resources defined (uses above CSI Driver)
+- Snapshot APIs enabled
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:17]: I think all these and third point  can be combined into
- VolumeSnapshots should be supported by CSI Driver and respective CRDs should be available

File: examples/stable/mysql-csi-snapshot/README.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+1 row in set (0.00 sec)
+```
+
+## Backup MySQL DB
+
+### Create Blueprint
+
+Create Blueprint in the same namespace as the Kanister controller.
+
+> **Note**: We used a Kubernetes cluster on DigitalOcean. Hence, snapshotClass and storageClass in the ./mysql-csi-snapshot-blueprint.yaml file is set to `do-block-storage`. Please correct these arguments as per your cluster setup. You can make use of `kubectl patch` or `kubectl edit` commands to make these changes.
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:91]:
```suggestion
> **Note**: We have set up the Kubernetes cluster on DigitalOcean. Hence, `snapshotClass` and `storageClass` in the `./mysql-csi-snapshot-blueprint.yaml` file is set to `do-block-storage`. Please correct these arguments as per your cluster setup before creating the blueprint.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+```bash
+$ kubectl create -f ./mysql-csi-snapshot-blueprint.yaml -n kanister
+```
+
+### Backup the application data
+
+Take a backup of the MySQL data using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:99]:
```suggestion
Snapshot the MySQL database using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+```bash
+# Create Actionset
+# Please make sure the value of blueprint matches with the name of blueprint that was created earlier
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:103]: let's explain `--pvc` thing here.

File: examples/stable/mysql-csi-snapshot/README.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+| mysql              |
+| performance_schema |
+| sys                |
++--------------------+
+5 rows in set (0.01 sec)
+```
+
+## Restore MySQL DB
+
+To restore the missing data, you should use the backup that you created before. We use `kanctl` for creating this restore action. `kanctl` helps create ActionSets that depend on other ActionSets.
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:162]:
```suggestion
To restore the missing data, you should use the backup that you created before. We will use `kanctl` for creating this restore action.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+# Use following command to debug if a new PVC in the format '<PVC-name>-restored' is not created
+$ kubectl --namespace kanister describe actionset restore-backup-mlvcv-6z9xn
+```
+
+Next, we need to add the newly restored PVC `data-mysql-release-0-restored` in the `mysql-release` StatefulSet spec template. One of the easiest way you can do that is by using `kubectl patch` command.
+
+```bash
+$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/volumes/-", "value": {"name": "restored-data", "persistentVolumeClaim": {"claimName": "data-mysql-release-0-restored"}}}]'
+
+$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/volumeMounts", "value": [{"mountPath": "/bitnami/mysql", "name": "restored-data"}, {"mountPath": "/opt/bitnami/mysql/conf/my.cnf", "name": "config", "subPath": "my.cnf"}]}]'
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:184]: should these be statefulsets or deployments.

File: examples/stable/mysql-csi-snapshot/README.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+```
+
+Once the patch is complete and MySQL pod is set to “Running“, you can check if the data has been restored to previous state.
+
+## Verify the restored application data
+
+To verify if restore was successful, you need to connect to the MySQL CLI and query the test database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:191]:
```suggestion
To verify if restore was successful, you need to connect to the MySQL CLI and query the `test` database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+mysql> SELECT * FROM employees;
++------+---------+-------------------------------+------+------------+-------------+
+| name | manager | address                       | sex  | birth      | department  |
++------+---------+-------------------------------+------+------------+-------------+
+| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
++------+---------+-------------------------------+------+------------+-------------+
+1 row in set (0.00 sec)
+```
+
+Thus, with the help of Kanister, we have successfully backed up our MySQL application data and recovered it after a disaster.
```
Comment (viveksinghggits) [inline examples/stable/mysql-csi-snapshot/README.md:226]:
```suggestion
Thus, with the help of Kanister, we have successfully backed up our MySQL application and recovered it after a disaster.
```

File: examples/stable/mysql-csi-snapshot/README.md (hunk 5 above)
Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/README.md:6]: How about saying "....CSI VolumeSnapshots to **capture** and restore MySQL database"? Using snapshot again doesn't sound good when you read it.

File: examples/stable/mysql-csi-snapshot/README.md (hunk 11 above)
Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/README.md:184]: These should be statefulsets

File: examples/stable/mysql-csi-snapshot/README.md (hunk 9 above)
Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/README.md:100]: Done

**Review by viveksinghggits — APPROVED** [review]

File: examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  name: mysql-csi-snapshot-bp
+actions:
+  backup:
+    outputArtifacts:
+      snapshotInfo:
+        keyValue:
+          name: "{{ .Phases.createCSISnapshot.Output.name }}"
+          pvc: "{{ .Phases.createCSISnapshot.Output.pvc }}"
+          namespace: "{{ .Phases.createCSISnapshot.Output.namespace }}"
+          restoreSize: "{{ .Phases.createCSISnapshot.Output.restoreSize }}"
```
Comment (oliverblaha) [inline examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: The restoreSize field from the VolumeSnapshot is allowed to be zero, if unknown. Even if it is set, it is supposed to be the minimum size required for restoring the actual data.

Depending on the CSI driver, this might therefore completely fail (if restoreSize is zero), or it is possible that the size of the resulting PVC will be just as much as is needed to restore the data. In that case, you have your data back, but probably can't do much with the new PVC unless you manually clone it into a bigger one.

Is it possible to either manually specify the restore size, or even track the original PVC's size somehow?

---

I also tested it with OpenEBS cstor - restoreSize remains zero, leading to an error:

> Warning  ActionSetFailed Action: restore  14s   Kanister Controller  Failed to execute phase: v1alpha1.Phase{Name:"restoreCSISnapshot", State:"pending", Output:map[string]interface {}(nil)}: PersistentVolumeClaim "demo-cstor-vol5-restored" is invalid: spec.resources[storage]: Invalid value: "0": must be greater than zero

---

References:

> restoreSize represents the complete size of the snapshot in bytes. In dynamic snapshot creation case, this field will be filled in by the CSI snapshotter sidecar with the "size_bytes" value returned from CSI "CreateSnapshot" gRPC call. For a pre-existing snapshot, this field will be filled with the "size_bytes" value returned from the CSI "ListSnapshots" gRPC call if the driver supports it. When restoring a volume from this snapshot, the size of the volume MUST NOT be smaller than the restoreSize if it is specified, otherwise the restoration will fail. If not specified, it indicates that the size is unknown.

Snapshot spec (protobuf message):

> // Information about a specific snapshot.
message Snapshot {
  // This is the complete size of the snapshot in bytes. The purpose of
  // this field is to give CO guidance on how much space is needed to
  // create a volume from this snapshot. The size of the volume MUST NOT
  // be less than the size of the source snapshot. This field is
  // OPTIONAL. If this field is not set, it indicates that this size is
  // unknown. The value of this field MUST NOT be negative and **a size of
  // zero means it is unspecified.**
  int64 size_bytes = 1;

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: @oliverblaha We do keep track of original PVC's restoreSize using the `outputArtifacts`. The backup action of blueprint records the restoreSize as an outputArtifact which is further used as an inputArtifact in restore action. So when we create a restore actionset from backup actionset, the kanister controller picks up the original PVC's restoreSize and allocates the same amount of memory to the newly restored PVC. You can refer [this](file:///home/infracloud/kanister/docs/_build/html/templates.html#artifacts) section in the docs to know more about Artifacts in Kanister.

You can also manually tweak restoreSize field in blueprint before running Restore steps. Just need to edit `.actions.restore.phases[0].args.restoreSize` to a suitable value in a format like 5Gi, 2Gi, etc.

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: @oliverblaha Can you share the blueprint you used to understand the issue?

Comment (oliverblaha) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: @shlokchaudhari9 I used the blueprint from this PR (just gave it a different name, that's all).
I believe that PVC does not have a `restoreSize` - so I assume you're referring to the snapshot's restoreSize field, am I correct? Unfortunately, that's the one that may be not set (= zero).

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: Yes. I am referring to the VolumeSnapshot's `restoreSize` field. The one mentioned in the Status of VolumeSnapshot. 

As far as I understand, `restoreSize` is a size_bytes status field returned by the [Snapshot Controller](https://github.com/kubernetes-csi/external-snapshotter). We can't set the value of` restoreSize` field to 0 and check for failure, as there is no way to mention `restoreSize` from the [Spec](https://github.com/kubernetes-csi/external-snapshotter/blob/ea7834929a37af7af5b63a979001cdddc0372f4e/client/apis/volumesnapshot/v1/types.go#L75).

In the blueprint, there are two `restoreSize` fields. Both of which have different meanings:
1. In the output of create snapshot function -This is the  `restoreSize` from the Status of VolumeSnapshot
2. In the input of restore snapshot function - This is the storage request to be made in the new PVC. 

The value in the second `restoreSize` field MUST be greater than or equal to the outputted `restoreSize` from create snapshot function.

The error log above is correct, since it was produced for 2nd case (PVC restore)

> I also tested it with OpenEBS cstor - restoreSize remains zero, leading to an error
Could you maybe help me with a detailed description of this test scenario? Mainly along the following lines.
1. Original PVC manifest
2. Steps used to perform backup action
3. Steps used to perform restore action
I am still curious about how you encountered this error without updating the example Blueprint from this PR. Maybe there's an edge-case we are not handling.

Comment (oliverblaha) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: IIRC, the field is zero by default, and then gets updated by the snapshot controller with the actual number of bytes as soon es the snapshot is ready, _if_ the ListSnapshots gRPC-call is supported and _if_ the size is known. Otherwise, it remains zero.

So I guess what happens here is simply OpenEBS not updating it, which seems to be perfectly legal.
Then the field remains zero, which is what you get back from the snapshotter. That's of course not very useful when trying to restore ;)

Is it possible to get the size of the PVC instead?

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: I see. Right now, we don't save the actual storage request the original PVC makes. We use the `restoreSize` from snapshot for the new PVC. So, the size of original PVC is something you might have to edit in the blueprint before trying to restore.
However, if this turns out to be a bug, one way to fix this would be by adding storage request size from original PVC in the `outputAritifacts` of create action.
But, before we jump to conclusions, I will definitely try out the OpenEBS cStor CSI driver for myself and update my findings here.

Comment (shlokchaudhari9) [reply examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml:13]: Hey @oliverblaha I have created an issue regarding above error that you had https://github.com/kanisterio/kanister/issues/1195
Let's track the progress regarding this use-case on the issue, hereafter.

PS: Could you please update the Kubernetes Version/Provider in environment details there? Also, please feel free to edit or add any information in the description :)

File: examples/stable/mysql-csi-snapshot/README.md [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+1 row in set (0.00 sec)
+```
+
+## Backup MySQL DB
+
+### Create Blueprint
+
+Create Blueprint in the same namespace as the Kanister controller.
+
+> **Note**:  We have set up the Kubernetes cluster on DigitalOcean. Hence, `snapshotClass` and `storageClass` in the `./mysql-csi-snapshot-blueprint.yaml` file is set to `do-block-storage`. Please correct these arguments as per your cluster setup before creating the blueprint.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:88]:
```suggestion
> **Note**:  This example uses a Kubernetes cluster on DigitalOcean. Therefore the `snapshotClass` and `storageClass` in the `./mysql-csi-snapshot-blueprint.yaml` file are set to `do-block-storage`. Change the arguments appropriately before creating the blueprint.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+# Install the MySQL database (Helm Version 3)
+$ kubectl create namespace mysql
+$ helm install mysql-release bitnami/mysql --namespace mysql \
+    --set auth.rootPassword='asd#45@mysqlEXAMPLE' \
+    --set architecture="standalone"
+```
+
+Above command deploys a MySQL instance in the `mysql` namespace.
+
+To retrieve your root password run the following command.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:36]:
```suggestion
Retrieve the root password with the following command.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+
+```bash
+$ kubectl get secret mysql-release --namespace mysql -o jsonpath="{.data.mysql-root-password}" | base64 --decode`
+```
+
+> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:42]:
```suggestion
> **Tip**: List all releases using `helm list --all-namespaces`.
```
since we already have helm3 as a pre-req

File: examples/stable/mysql-csi-snapshot/README.md [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
++------+---------+-------------------------------+------+------------+-------------+
+1 row in set (0.00 sec)
+```
+
+## Backup MySQL DB
+
+### Create Blueprint
+
+Create Blueprint in the same namespace as the Kanister controller.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:86]:
```suggestion
Create the Blueprint in the `kanister` namespace.
```
Some instructions directly say this

File: examples/stable/mysql-csi-snapshot/README.md [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+$ kubectl create -f ./mysql-csi-snapshot-blueprint.yaml -n kanister
+```
+
+### Backup the application data
+
+Snapshot MySQL database using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.
+
+```bash
+# Create Actionset
+# Please make sure the value of blueprint matches with the name of blueprint that was created earlier
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:100]:
```suggestion
# Make sure the value of blueprint matches the name of blueprint created earlier
```
Second inconsistency noticed here and in other docs as well. We should either use `Please` everywhere or not use at all.

File: examples/stable/mysql-csi-snapshot/README.md [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+| mysql              |
+| performance_schema |
+| sys                |
++--------------------+
+5 rows in set (0.01 sec)
+```
+
+## Restore MySQL DB
+
+To restore the missing data, you should use the backup that you created before. We will use `kanctl` for creating this restore action.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:160]:
```suggestion
To restore the missing data, use the backup created earlier. This can be achieved using `kanctl` for creating the restore action.
```
Noticed this inconsistency in all the other docs as well. Something to do in the future for others but let's start with this one. We should either not use `you, your etc.` or `we, us, our etc.` anywhere in the document or use it everywhere. I would recommend making the documentation neutral.
For example:
One of the sentences above says
`Create the Blueprint in the kanister namespace.` which is neutral instead of saying, `You can create a blueprint in the kanister namespace` which is directed towards the reader.

File: examples/stable/mysql-csi-snapshot/README.md [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+# --pvc argument specifies the PVC on which the action is performed
+$ kanctl create actionset --action backup --namespace kanister --blueprint mysql-csi-snapshot-bp --pvc mysql/$(kubectl get pvc -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}')
+actionset backup-mlvcv created
+
+$ kubectl --namespace kanister get actionset
+NAME                         AGE
+backup-mlvcv                 112s
+
+# View the status of the actionset
+# Please make sure the name of the actionset here matches with name of the actionset created above
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:110]:
```suggestion
# Make sure the name of the actionset here matches the name of the actionset created above
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+## Restore MySQL DB
+
+To restore the missing data, you should use the backup that you created before. We will use `kanctl` for creating this restore action.
+
+```bash
+# Make sure to use correct backup actionset name here
+$ kanctl --namespace kanister create actionset --action restore --from backup-mlvcv
+actionset restore-backup-mlvcv-6z9xn created
+
+# Check the PVCs in mysql namespace. You should see a new PVC in the format '<PVC-name>-restored'
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:167]:
```suggestion
# Check the PVCs in mysql namespace. A new PVC is created with the name of format '<PVC-name>-restored'
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+$ kubectl -n mysql  get pvc
+NAME                            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
+data-mysql-release-0            Bound    pvc-b966c5ef-3d20-4ec2-9ab3-1a868737ea0a   8Gi        RWO            do-block-storage   4m14s
+data-mysql-release-0-restored   Bound    pvc-2eb6fb89-78bc-4c2e-8c1e-c270d664f311   8Gi        RWO            do-block-storage   18s
+
+# Use following command to debug if a new PVC in the format '<PVC-name>-restored' is not created
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:173]:
```suggestion
# In case a new PVC is not created, use the following command to check events in the actionset
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 24]
```diff
@@ -0,0 +1,267 @@
...
+$ kubectl --namespace kanister describe actionset restore-backup-mlvcv-6z9xn
+```
+
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:176]: Not very confident with this approach. Here's what I think
1. Somewhere above, we should probably add a note, saying that this process may not work if application-consistent backups are required.
2. Recommend scaling down the workload (SS) deleting the existing PVC and restoring one with the same name above, and finally scaling up the workload to achieve a consistent restore.

File: examples/stable/mysql-csi-snapshot/README.md [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/volumes/-", "value": {"name": "restored-data", "persistentVolumeClaim": {"claimName": "data-mysql-release-0-restored"}}}]'
+
+$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/volumeMounts", "value": [{"mountPath": "/bitnami/mysql", "name": "restored-data"}, {"mountPath": "/opt/bitnami/mysql/conf/my.cnf", "name": "config", "subPath": "my.cnf"}]}]'
+```
+
+Once the patch is complete and MySQL pod is set to “Running“, you can check if the data has been restored to previous state.
+
+## Verify the restored application data
+
+To verify if restore was successful, you need to connect to the MySQL CLI and query the `test` database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:189]:
```suggestion
To verify if the restore was successful, connect to the MySQL CLI and query the `test` database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+mysql> SELECT * FROM employees;
++------+---------+-------------------------------+------+------------+-------------+
+| name | manager | address                       | sex  | birth      | department  |
++------+---------+-------------------------------+------+------------+-------------+
+| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
++------+---------+-------------------------------+------+------------+-------------+
+1 row in set (0.00 sec)
+```
+
+Thus, with the help of Kanister, we have successfully backed up our MySQL application and recovered it after a disaster.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:224]:
```suggestion
```
Not required

File: examples/stable/mysql-csi-snapshot/README.md [hunk 27]
```diff
@@ -0,0 +1,267 @@
...
+Install the MySQL database using the bitnami chart with the release name `mysql-release`.
+
+```bash
+# Add bitnami in your local chart repository
+$ helm repo add bitnami https://charts.bitnami.com/bitnami
+
+# Update your local chart repository
+$ helm repo update
+
+# Install the MySQL database (Helm Version 3)
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:27]:
```suggestion
# Install the MySQL database
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 28]
```diff
@@ -0,0 +1,267 @@
...
+```
+
+## Cleanup
+
+### Uninstalling the helm chart
+
+To uninstall/delete the `mysql-release` deployment.
+
+```bash
+# Helm Version 3
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:246]:
```suggestion
# Delete the MySQL database
```

File: examples/stable/mysql-csi-snapshot/README.md [hunk 29]
```diff
@@ -0,0 +1,265 @@
+# Backup and restore MySQL DB using CSI VolumeSnapshot
+
+VolumeSnapshots provide Kubernetes users with a standardized way to copy a volume's contents at a particular point in time without creating an entirely new volume. This functionality enables, for example, database administrators to backup databases before performing edit or delete modifications.
+
+## Introduction
+This document explains how Kanister leverages the use of CSI VolumeSnapshots to capture and restore MySQL database.
```
Comment (pavannd1) [inline examples/stable/mysql-csi-snapshot/README.md:6]: As discussed offline, let's use a non-DB application for this example. Generally, I assume that the blueprint would be used directly on PVCs but if we want to talk about an application, let's use timelogger.

**Review by pavannd1 — CHANGES_REQUESTED** [review]
As discussed offline, let's change this to use a different application like timelogger.

**Comment by shlokchaudhari9** [comment]
> As discussed offline, let's change this to use a different application like timelogger.

@pavannd1 This PR is ready to be reviewed. Please have a look.

File: examples/stable/csi-snapshot/README.md [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+
+```bash
+# Run a shell inside time-logger deployment's pod
+$ kubectl exec -it $(kubectl get pods -n time-log -l app=time-logger -o=jsonpath='{.items[0].metadata.name}') -n time-log -- /bin/bash
+
+# Remove the log file
+$ rm /var/log/time.log
+
+# Check the first entry recorded in the /var/log/time.log again. It is now replaced with new entry.
+# This is because Time Log app recreates the log file once it's deleted and starts adding newer entries to it.
```
Comment (pavannd1) [inline examples/stable/csi-snapshot/README.md:83]:
```suggestion
# This is because Time-Logger app recreates the log file once it's deleted and starts adding newer entries to it.
```

File: examples/stable/csi-snapshot/README.md [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+Let's say someone accidentally deleted the log file at path `/var/log`
+
+```bash
+# Run a shell inside time-logger deployment's pod
+$ kubectl exec -it $(kubectl get pods -n time-log -l app=time-logger -o=jsonpath='{.items[0].metadata.name}') -n time-log -- /bin/bash
+
+# Remove the log file
+$ rm /var/log/time.log
+
+# Check the first entry recorded in the /var/log/time.log again. It is now replaced with new entry.
```
Comment (pavannd1) [inline examples/stable/csi-snapshot/README.md:82]:
```suggestion
# Check the first entry recorded in the /var/log/time.log again. It is now replaced with a new entry.
```

File: examples/stable/csi-snapshot/README.md [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+$ kanctl --namespace kanister create actionset --action restore --from backup-mlvcv
+actionset restore-backup-mlvcv-6z9xn created
+
+# Use the following command to check events in the actionset
+$ kubectl --namespace kanister describe actionset restore-backup-mlvcv-6z9xn
+```
+
+## Verify the restored application data
+
+To verify restore action, we check the first entry of file `/var/log/time.log` and confirm that it's been restored with the original value
```
Comment (pavannd1) [inline examples/stable/csi-snapshot/README.md:103]:
```suggestion
To verify restore action, check the first entry of the file `/var/log/time.log` and confirm that it's been restored with the original value.
```

**Review by pavannd1 — APPROVED** [review]
LGTM
